### PR TITLE
POC: add telemetry to localprocess backend.

### PR DESCRIPTION
**Pull request by sujalshah-bit**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** _(optional, in `Fixes #<issue number>, #<issue number>, ...` format, will close the issue(s) when PR gets merged)_:

Fixes #

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/) included if any changes are user facing

### CODE REVIEW

**Comment by sujalshah-bit** [comment]
<p><strong>POC: Kubeflow Trainer + OpenTelemetry — two pain points demonstrated on KIND (CPU-only)</strong></p>
<p><strong>What's implemented:</strong> OTel spans across the full chain — <code>TrainerClient.train()</code> → K8s backend → worker pods — with W3C <code>traceparent</code> injected into the TrainJob CRD env vars so all worker spans share one <code>traceID</code>.</p>
<hr>
<p><strong>Pain point 1 — Straggler worker</strong></p>
<blockquote>
  <p><em>"When something fails, figuring out what actually went wrong takes too long." — Kubeflow survey</em></p>
</blockquote>
<p>Scattered logs across pods make it hard to identify which worker is slow. Logs only print loss <em>after</em> <code>dist.barrier()</code> — when both ranks have already synced — so per-rank per-epoch duration never appears anywhere. You're left manually diffing timestamps across two pod log streams.</p>
<p>The trace has <code>training.epoch_duration_ms</code> set <em>before</em> the barrier, per rank, per epoch:</p>

  | rank 0 | rank 1 (straggler)
-- | -- | --
epoch 1 | 141ms | 341ms
epoch 2 | 102ms | 301ms
epoch 3 | 105ms | 304ms

localhost_16686_trace_8180399a12c3b0f388817a7fc1842e6e (1)

<p>One trace view. No log diffing.</p>
<hr>
<p><strong>Pain point 2 — Misdiagnosed failure</strong></p>
<blockquote>
<p><em>"Debugging distributed jobs running on Kubernetes remains a major challenge... cryptic errors such as... networking issues during distributed training."— Kubeflow survey</em></p>
</blockquote>
<p>When rank 0 crashes, rank 1 dies with <code>Connection closed by peer [gloo/tcp]</code> — which looks exactly like a networking issue. Wrong diagnosis, wasted time.</p>
<p>From the trace: both workers are siblings under the same parent span on a shared microsecond timeline. Rank 0's <code>training.epoch [2]</code> is marked <code>ERROR</code> with <code>training.loss = 0.9238</code>. Rank 1's gloo error fires 325ms later as a cascade. Cause vs consequence is in the data — no cross-referencing two log files.</p>
<hr>

https://github.com/user-attachments/assets/f7a39598-0811-4263-808b-c213fd0958ce

**Notebook**: https://github.com/sujalshah-bit/kf-otel-demo/blob/main/kubeflow-otel-showcase.ipynb

_source:
- [Kubeflow survery](https://docs.google.com/document/d/1PFDM6-AxliRykfyVOcRs3mT6N8qmIC2z-9Fj2Lrbwlo/edit?tab=t.0)
- [To identify pain points, I scraped and analyzed data from various online sources.](https://sujalshah-bit.github.io/need-for-otel-integration-kf.github.io/)
